homeassistant/components/mobile_app/__init__.py

```diff
@@ -1,38 +1,120 @@
+from homeassistant.helpers.discovery import load_platform
 from homeassistant.helpers.typing import ConfigType, HomeAssistantType
 
-from .const import (DATA_DELETED_IDS, DATA_REGISTRATIONS, DATA_STORE, DOMAIN,
-                    STORAGE_KEY, STORAGE_VERSION)
+from .const import (ATTR_APP_COMPONENT, ATTR_DEVICE_ID, ATTR_DEVICE_NAME,
+                    ATTR_MANUFACTURER, ATTR_MODEL, ATTR_OS_VERSION,
+                    DATA_CONFIG_ENTRIES, DATA_DELETED_IDS, DATA_DEVICES,
+                    DATA_STORE, DOMAIN, STORAGE_KEY, STORAGE_VERSION)
 
-from .http_api import register_http_handlers
-from .webhook import register_deleted_webhooks, setup_registration
+from .http_api import RegistrationsView
+from .webhook import handle_webhook
 from .websocket_api import register_websocket_handlers
 
 DEPENDENCIES = ['device_tracker', 'http', 'webhook']
 
 REQUIREMENTS = ['PyNaCl==1.3.0']
 
+_LOGGER = logging.getLogger(__name__)
+
 
 async def async_setup(hass: HomeAssistantType, config: ConfigType):
     """Set up the mobile app component."""
+    hass.data[DOMAIN] = {
+        DATA_CONFIG_ENTRIES: {}, DATA_DELETED_IDS: [], DATA_DEVICES: {},
+    }
+
     store = hass.helpers.storage.Store(STORAGE_VERSION, STORAGE_KEY)
     app_config = await store.async_load()
     if app_config is None:
-        app_config = {DATA_DELETED_IDS: [], DATA_REGISTRATIONS: {}}
-
-    if hass.data.get(DOMAIN) is None:
-        hass.data[DOMAIN] = {DATA_DELETED_IDS: [], DATA_REGISTRATIONS: {}}
+        app_config = {
+            DATA_CONFIG_ENTRIES: {}, DATA_DELETED_IDS: [], DATA_DEVICES: {},
+        }
 
-    hass.data[DOMAIN][DATA_DELETED_IDS] = app_config.get(DATA_DELETED_IDS, [])
-    hass.data[DOMAIN][DATA_REGISTRATIONS] = app_config.get(DATA_REGISTRATIONS,
-                                                           {})
+    hass.data[DOMAIN] = app_config
     hass.data[DOMAIN][DATA_STORE] = store
 
-    for registration in app_config[DATA_REGISTRATIONS].values():
-        setup_registration(hass, store, registration)
-
-    register_http_handlers(hass, store)
+    hass.http.register_view(RegistrationsView())
     register_websocket_handlers(hass)
-    register_deleted_webhooks(hass, store)
+
+    for deleted_id in hass.data[DOMAIN][DATA_DELETED_IDS]:
+        try:
+            webhook_register(hass, DOMAIN, "Deleted Webhook", deleted_id,
+                             handle_webhook)
+        except ValueError:
+            pass
+
+    return True
+
+
+async def async_setup_entry(hass, entry):
+    """Set up a mobile_app entry."""
+    registration = entry.data
+
+    webhook_id = registration[CONF_WEBHOOK_ID]
+
+    hass.data[DOMAIN][DATA_CONFIG_ENTRIES][webhook_id] = entry
+
+    device_registry = await dr.async_get_registry(hass)
+
+    identifiers = {
+        (ATTR_DEVICE_ID, registration[ATTR_DEVICE_ID]),
+        (CONF_WEBHOOK_ID, registration[CONF_WEBHOOK_ID])
+    }
+
+    config_entry_id = entry.entry_id
+
+    if config_entry_id is None:
```

Comment:
This cannot happen, you literally just got this passed in.

Reply:
tru